proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+and/or clean up state when an exception is thrown. Like other higher-level
+constructs, a try block begins with a `try` instruction, and ends with an `end`
+instruction. That is, a try block is sequence of instructions having the
+following form:
+
+```
+try block_type
+  instruction*
+catch
+  instruction*
+end
+```
+
+A try block ends with a `catch block` that is defined by the list of
+instructions after the `catch` instruction.
+
+Try blocks, like control-flow blocks, have a _block type_. The block type of a
+try block defines the values yielded by the evaluation the try block when either
+no exception is thrown, or the exception is successfully caught by the catch
+block.
+
+In the initial implementation, try blocks may only yield 0 or 1 values.
+
+### Exception creation
+
+A `except` instruction has a single immediate argument, an exception tag.  The
+corresponding tagged exception type is used to define the data fields of the
+created exception. The values for the data fields must be on top of the operand
+stack, and must correspond to the exception's type signature.  These values are
+popped off the stack and an instance of the exception is then created. A
+reference to the created exception is then pushed onto the stack.
+
+### Throws
+
+The `throw` throws the exception on top of the stack. The exception is popped
+off the top of the stack before throwing.
+
+When an exception is thrown, the host VM searches for nearest enclosing try
+block body that execution is in. That try block is called the _catching_ try
+block.
+
+If the throw appears within the body of a try block, it is the catching try
+block.
+
+If a throw occurs within a function body, and it doesn't appear inside the body
+of a try block, the throw continues up the call stack until it is in the body of
+an an enclosing try block, or the call stack is flushed. If the call stack is
+flushed, the host VM defines how to handle uncaught exceptions.  Otherwise, the
+found enclosing try block is the catching try block.
+
+A throw inside the body of a catch block is never caught by the corresponding
+try block of the catch block, since instructions in the body of the catch block
+are not in the body of the try block.
+
+Once a catching try block is found for the throw, the operand stack is popped back
+to the size the operand stack had when the try block was entered, and then
+the caught exception is pushed onto the stack.
+
+If control is transferred to the body of a catch block, and the last instruction
+in the body is executed, control then exits the try block.
+
+If the selected catch block does not throw an exception, it must yield the
+value(s) expected by the corresponding catching try block. This includes popping
+the caught exception.
+
+Note that a caught exception can be rethrown using the `throw` instruction.
+
+### Exception data extraction
+
+The `if_except block` defines a conditional query of the exception on top of the
```

Comment:
This adds another type of control-flow block specifically for this purpose, which has an implementation cost. What alternatives have been considered, e.g. an instruction that loads the type or Nth element from an exception object, such that a the existing if block can be used to inspect it?

Reply:
I agree that there is an implementation cost to the current proposal. On the other hand, there
may be implementation cost of loading elements individually - I.e. does the element loader need to verify the access is correct on each element lookup.

In the current vacuum of realistic examples, it is hard to say which is better, so one was chosen, based on previous discussions in the original proposal (in Exceptions.md).

Reply:
I think the main advantage of the approach we've presented here is the guarantees provided by the type system. Let's say we had a `get_type_index` instruction and a `get_exception_field` instruction, we'd end up with code like:

```
if get_type_index == 0:
  get_exception_field 0, 0
else:
  ...
```

The problem is that at `get_exception_field`, we don't statically know the exception has the correct type (here it's pretty obvious because it shows up immediately after a type check, but it may not always be that easy to see).  This means we have to insert a dynamic type check on top of the one we just generated.

Instead, the combined match and unpack that we have with `if_except` eliminates the need to insert any dynamic check beyond what the programmer has written already.